core/src/main/java/org/elasticsearch/plugins/PluginManager.java

```diff
@@ -162,97 +165,44 @@ public void downloadAndExtract(String name, Terminal terminal) throws IOExceptio
         }
 
         if (!downloaded) {
+            // try to cleanup what we downloaded
+            IOUtils.deleteFilesIgnoringExceptions(pluginFile);
             throw new IOException("failed to download out of all possible locations..., use --verbose to get detailed information");
         }
+        return pluginFile;
+    }
 
-        // unzip plugin to a temp dir
-        Path tmp = unzipToTemporary(pluginFile);
-
-        // create list of current jars in classpath
-        final List<URL> jars = new ArrayList<>();
-        ClassLoader loader = PluginManager.class.getClassLoader();
-        if (loader instanceof URLClassLoader) {
-            Collections.addAll(jars, ((URLClassLoader) loader).getURLs());
+    private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFile) throws IOException {
+        final Path extractLocation = pluginHandle.extractedDir(environment);
+        if (Files.exists(extractLocation)) {
+            throw new IOException("plugin directory " + extractLocation.toAbsolutePath() + " already exists. To update the plugin, uninstall it first using remove " + pluginHandle.name + " command");
```

Comment:
i am not even sure if that is really the command to recommend, since pluginName is "modified" (e.g. shield not elasticesearch-shield) by this logic. I will look into it...

This statement and others i just moved around (refactored out of huge method into smaller methods).